File: code/python/ch17/practice/warmup_02_heap_sort.py
"""
Warmup 2: Sort Using Heap (Heapsort)
========================================
Chapter 17: Heaps & Priority Queues — The VIP Line

PROBLEM
-------
Given an unsorted array of integers, return the array sorted in ascending order
using a heap-based approach (heapsort).

INPUT FORMAT
------------
A single line of space-separated integers.

OUTPUT FORMAT
-------------
The sorted array as a list.

CONSTRAINTS
-----------
- 0 <= len(arr) <= 10^5
- -10^9 <= arr[i] <= 10^9

EXAMPLES
--------
Input:
  5 3 8 1 2
Output: [1, 2, 3, 5, 8]

Input:
  1
Output: [1]

HINT
----
Push all elements into a min-heap, then pop them all out.
They come out in sorted order!

INSTRUCTIONS
------------
Replace the `pass` in the solve() function with your solution.
The main block at the bottom handles input/output — don't change it.
"""


def solve(arr: list[int]) -> list[int]:
    """Return arr sorted in ascending order using a heap."""
    pass  # TODO: Replace this with your solution


# ── Do not change anything below this line ──────────────────────────
if __name__ == "__main__":
    line = input().strip()
    if line:
        arr = list(map(int, line.split()))
    else:
        arr = []
    print(solve(arr))

File: code/cpp/ch17/practice/warmup_02_heap_sort.cpp
/*
 * Warmup 2: Sort Using Heap (Heapsort)
 * ======================================
 * Chapter 17: Heaps & Priority Queues — The VIP Line
 *
 * PROBLEM:
 *   Given an array of integers, return it sorted in ascending order
 *   using a heap (priority queue).
 *
 * EXAMPLES:
 *   solve({5,3,8,1,2})  -> {1,2,3,5,8}
 *   solve({})            -> {}
 *
 * CONSTRAINTS:
 *   - 0 <= arr.size() <= 10^5
 *   - Elements can be any integer
 *
 * INSTRUCTIONS:
 *   Replace the body of solve() with your solution.
 */

#include <algorithm>
#include <functional>
#include <iostream>
#include <queue>
#include <vector>
using namespace std;

vector<int> solve(vector<int> arr) {
    // TODO: Replace this with your solution
    return {};
}

// -- Do not change anything below this line --------------------------
int main() {
    int n;
    cin >> n;
    vector<int> arr(n);
    for (int i = 0; i < n; i++) cin >> arr[i];
    vector<int> result = solve(arr);
    for (int i = 0; i < (int)result.size(); i++) {
        if (i) cout << " ";
        cout << result[i];
    }
    cout << endl;
    return 0;
}
